pkg/apiserver/api_installer.go

```diff
@@ -0,0 +1,246 @@
+	// Note: update GetAttribs() when adding a handler.
+	// Watch.
+	if _, ok := storage.(ResourceWatcher); ok {
+		// Watch is supported on this resource.
+		watchKindRoute := ws.GET("/watch/" + path).To(restfulStripPrefix(a.prefix+"/watch", watchHandler)).
+			Doc("watch " + kind).
+			Operation("watch" + kind)
+		addParamIf(watchKindRoute, namespaceParam, namespaceScope)
+		ws.Route(watchKindRoute)
+
+		watchNamedRoute := ws.GET("/watch/" + path + "/{name}").To(restfulStripPrefix(a.prefix+"/watch", watchHandler)).
+			Doc("watch a particular " + kind).
+			Operation("watchNamed" + kind).
+			Param(nameParam)
+		addParamIf(watchNamedRoute, namespaceParam, namespaceScope)
+		ws.Route(watchNamedRoute)
+	}
+
+	// Redirect and proxy
+	if _, ok := storage.(Redirector); ok {
+		// Redirect and proxy are supported on this resource.
+		redirectRoute := ws.GET("/redirect/" + path + "/{name}").To(restfulStripPrefix(a.prefix+"/redirect", redirectHandler)).
+			Doc("get the redirect URL for " + kind).
+			Operation("redirect" + kind).
+			Param(nameParam)
+		addParamIf(redirectRoute, namespaceParam, namespaceScope)
+		ws.Route(redirectRoute)
+
+		// Accept all methods as per https://github.com/GoogleCloudPlatform/kubernetes/issues/3996
+		addProxyRoute(ws, "GET", a.prefix, path, proxyHandler, kind, nameParam, namespaceParam, namespaceScope)
+		addProxyRoute(ws, "PUT", a.prefix, path, proxyHandler, kind, nameParam, namespaceParam, namespaceScope)
+		addProxyRoute(ws, "POST", a.prefix, path, proxyHandler, kind, nameParam, namespaceParam, namespaceScope)
+		addProxyRoute(ws, "DELETE", a.prefix, path, proxyHandler, kind, nameParam, namespaceParam, namespaceScope)
+	}
+
+	return nil
+}
+
+// This magic incantation returns *ptrToObject for an arbitrary pointer
+func indirectArbitraryPointer(ptrToObject interface{}) interface{} {
+	return reflect.Indirect(reflect.ValueOf(ptrToObject)).Interface()
+}
+
+// Adds the given param to the given route builder if shouldAdd is true. Does nothing if shouldAdd is false.
+func addParamIf(b *restful.RouteBuilder, parameter *restful.Parameter, shouldAdd bool) *restful.RouteBuilder {
+	if !shouldAdd {
+		return b
+	}
+	return b.Param(parameter)
+}
+
+// Returns a restful RouteFunction that calls the given handler after stripping prefix from the request path.
+func restfulStripPrefix(prefix string, handler http.Handler) restful.RouteFunction {
+	return func(restReq *restful.Request, restResp *restful.Response) {
+		//	handler.ServeHTTP(restResp.ResponseWriter, restReq.Request)
+		http.StripPrefix(prefix, handler).ServeHTTP(restResp.ResponseWriter, restReq.Request)
+	}
+}
+
+func addProxyRoute(ws *restful.WebService, method string, prefix string, path string, proxyHandler http.Handler, kind string, nameParam *restful.Parameter, namespaceParam *restful.Parameter, namespaceScope bool) {
+	proxyRoute := ws.Method(method).Path("/proxy/" + path + "/{name}/{*}").To(restfulStripPrefix(prefix+"/proxy", proxyHandler)).
```

Comment:
What does the swagger for this look like? Please add an identifier to the trailing subpath, such as `{path:*}`.

Are you sure that the default router doesn't work for such paths? I know the test and examples use the curly router, but it looks like the path tokenization is actually in common code rather than in the routers.
https://github.com/emicklei/go-restful/blob/master/route.go#L113

I've come to the conclusion that proxy and redirect aren't really part of the API, but are convenience functions provided by the apiserver. However, since the do need to be registered on a per-object basis, and we do need some kind of documentation for them, I can see the rationale for registering them with the other handlers.